pkg/signup/signup_client.go

```diff
@@ -0,0 +1,114 @@
+package signup
+
+import (
+	crtapi "github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/rest"
+	// See note below
+	//v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	USER_SIGNUP_RESOURCE_NAME = "usersignups"
+)
+
+func NewUserSignupClient(cfg *rest.Config) (*UserSignupV1Alpha1Client, error) {
+	scheme := runtime.NewScheme()
+	err := crtapi.SchemeBuilder.AddToScheme(scheme)
+	if err != nil {
+		return nil, err
+	}
+	crtapi.SchemeBuilder.Register(&crtapi.UserSignup{}, &crtapi.UserSignupList{})
+
+	config := *cfg
+	config.GroupVersion = &crtapi.SchemeGroupVersion
+	config.APIPath = "/apis"
+	config.ContentType = runtime.ContentTypeJSON
+	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: serializer.NewCodecFactory(scheme)}
+	config.UserAgent = rest.DefaultKubernetesUserAgent()
+
+	client, err := rest.RESTClientFor(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &UserSignupV1Alpha1Client{
+		restClient: client,
+	}, nil
+}
+
+type UserSignupV1Alpha1Client struct {
```

Comment:
I followed the pattern as recommended by [this article](https://medium.com/velotio-perspectives/extending-kubernetes-apis-with-custom-resource-definitions-crds-139c99ed3477).  I believe it benefits us with some future-proofing as the actual client implementation is tied to the CRD version.

I'm not sure we can easily introduce a generic client.. this one is specific to `UserSignup`, from the function parameters to its return types.  How do you think a generic client would look?

Reply:
While the article you mention is about something else I think I got the idea behind the versioning in the client names. It could help when we need to support multiple API versions during operator API updates.

So, I'm OK to keep it as is.

About the generic client... Yeah.. it's probably too tricky/not worth doing. But at least maybe we could create some helper function to create a rest client for the scheme and GV and reuse it in the MUR and UserSignup clients: 

```
func RestClient(cfg *rest.Config, gv schema.GroupVersion, scheme *runtime.Scheme) (rest.Interface, error) {
	config := *cfg
	config.GroupVersion = &gv
	config.APIPath = "/apis"
	config.ContentType = runtime.ContentTypeJSON
	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: serializer.NewCodecFactory(scheme)}
	config.UserAgent = rest.DefaultKubernetesUserAgent()

	client, err := rest.RESTClientFor(&config)
	if err != nil {
		return nil, err
	}

	return client, nil
}
```

Reply:
Or with some generic client:

```
import (
	crtapi "github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"

	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/apimachinery/pkg/runtime/serializer"
	"k8s.io/client-go/rest"
)

func NewV1Alpha1Client(cfg *rest.Config, namespace string, registerObjects ...runtime.Object) (*V1Alpha1Client, error) {
	scheme := runtime.NewScheme()
	err := crtapi.SchemeBuilder.AddToScheme(scheme)
	if err != nil {
		return nil, err
	}
	crtapi.SchemeBuilder.Register(registerObjects...)

	config := *cfg
	config.GroupVersion = &crtapi.SchemeGroupVersion
	config.APIPath = "/apis"
	config.ContentType = runtime.ContentTypeJSON
	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: serializer.NewCodecFactory(scheme)}
	config.UserAgent = rest.DefaultKubernetesUserAgent()

	client, err := rest.RESTClientFor(&config)
	if err != nil {
		return nil, err
	}

	return &V1Alpha1Client{
		restClient: client,
		ns:         namespace,
	}, nil
}

type V1Alpha1Client struct {
	restClient rest.Interface
	ns         string
}
```

Then UserSignup client impl:

```
const (
	userSignupResourcePlural = "usersignups"
)

func NewUserSignupV1Alpha1Client(cfg *rest.Config, namespace string) (*UserSignupV1Alpha1Client, error) {
	client, err := NewV1Alpha1Client(cfg, namespace, &crtapi.UserSignup{}, &crtapi.UserSignupList{})
	if err != nil {
		return nil, err
	}

	return &UserSignupV1Alpha1Client{client}, nil
}

type UserSignupV1Alpha1Client struct {
	*V1Alpha1Client
}

func (c *UserSignupV1Alpha1Client) Get(name string) (*crtapi.UserSignup, error) {
	result := &crtapi.UserSignup{}
	err := c.restClient.Get().
		Namespace(c.ns).
		Resource(userSignupResourcePlural).
		Name(name).
		Do().
		Into(result)
	return result, err
}

func (c *UserSignupV1Alpha1Client) Create(obj *crtapi.UserSignup) (*crtapi.UserSignup, error) {
	result := &crtapi.UserSignup{}
	err := c.V1Alpha1Client.restClient.Post().
		Namespace(c.ns).
		Resource(userSignupResourcePlural).
		Body(obj).
		Do().
		Into(result)
	return result, err
}
```

Reply:
And usage would be:

```
client, err := signup.NewUserSignupV1Alpha1Client(config, namespace)`
userSignUp, err := client.Get(name)
```
Or even `signup.NewV1Alpha1Client(config, namespace)` instead of `signup.NewUserSignupV1Alpha1Client(config, namespace)`.

WDYT?

Reply:
Resolving this as we discussed offline